umap/tests/test_datalayer_views.py

```diff
@@ -392,62 +393,157 @@ def test_anonymous_user_can_edit_if_inherit_and_map_in_public_mode(
+        url, post_data, follow=True, HTTP_IF_UNMODIFIED_SINCE=reference_timestamp
     )
     assert response.status_code == 200
     modified_datalayer = DataLayer.objects.get(pk=datalayer.pk)
-    assert modified_datalayer.geojson.read().decode() == (
-        '{"type": "FeatureCollection", "features": [{"type": "Feature", "geometry": {'
-        '"type": "Point", "coordinates": [-1, 2]}, "properties": {"_umap_options": {}'
-        ', "name": "foo"}}, {"type": "Feature", "geometry": {"type": "LineString", "c'
-        'oordinates": [2, 3]}, "properties": {"_umap_options": {}, "name": "bar"}}, {'
-        '"type": "Feature", "geometry": {"type": "Point", "coordinates": [3, 4]}, "pr'
-        'operties": {"_umap_options": {}, "name": "marker"}}, {"type": "Feature", "ge'
-        'ometry": {"type": "Point", "coordinates": [5, 6]}, "properties": {"_umap_opt'
-        'ions": {}, "name": "new from someone else"}}, {"type": "Feature", "geometry"'
-        ': {"type": "Point", "coordinates": [7, 8]}, "properties": {"_umap_options": '
-        '{}, "name": "new from us"}}], "_umap_options": {"displayOnLoad": true, "name'
-        '": "new name", "id": 1668, "remoteData": {}, "color": "LightSeaGreen", "desc'
-        'ription": "test"}}'
+    merged_features = json.load(modified_datalayer.geojson)["features"]
+
+    for reference_feature in reference_data["features"]:
+        assert reference_feature in merged_features
+
+    assert client1_feature in merged_features
+    assert client2_feature in merged_features
+
+
+def test_optimistic_merge_conflicting_change_raises(
+    client, datalayer, map, reference_data
+):
+    url = reverse("datalayer_update", args=(map.pk, datalayer.pk))
+    client.login(username=map.owner.username, password="123123")
+
+    # Reference data is:
+    # Point (-1, 2); Linestring (2, 3); Point (3, 4).
+
+    post_data = {
+        "name": "name",
+        "display_on_load": True,
+        "rank": 0,
+        "geojson": SimpleUploadedFile(
+            "foo.json", json.dumps(reference_data).encode("utf-8")
+        ),
+    }
+    response = client.post(url, post_data, follow=True)
+    assert response.status_code == 200
+
+    response = client.get(reverse("datalayer_view", args=(map.pk, datalayer.pk)))
+    reference_timestamp = response["Last-Modified"]
+
+    # The first client changes the first feature.
+    client1_data = deepcopy(reference_data)
+    client1_data["features"][0]["geometry"] = {"type": "Point", "coordinates": [5, 6]}
+
+    # Sleep to change the current timestamp (used in the If-Unmodified-Since header)
+    time.sleep(1)
+    post_data["geojson"] = SimpleUploadedFile(
+        "foo.json",
+        json.dumps(client1_data).encode("utf-8"),
     )
+    response = client.post(
+        url, post_data, follow=True, HTTP_IF_UNMODIFIED_SINCE=reference_timestamp
+    )
+    assert response.status_code == 200
+
+    # The first client changes the first feature.
```

Comment:
second ?

Reply:
Nope, that's the first. We don't actually care who sent the first base doc, so it's out of the picture.

Reply:
Humm, but how is the var name `client2_data` then ? :upside_down_face:

Reply:
Ah, that's the second one 😅 (sorry was missing context)